hermes/src/price_feeds_metadata.rs

```diff
@@ -0,0 +1,83 @@
+use {
+    crate::{
+        api::types::{
+            AssetType,
+            PriceFeedMetadata,
+        },
+        network::pythnet::fetch_and_store_price_feeds_metadata,
+        state::{
+            cache::AggregateCache,
+            retrieve_price_feeds_metadata,
+        },
+    },
+    anyhow::Result,
+};
+
+pub async fn get_price_feeds_metadata<S>(
+    state: &S,
+    query: Option<String>,
+    asset_type: Option<AssetType>,
+) -> Result<Vec<PriceFeedMetadata>>
+where
+    S: AggregateCache,
+    S: PriceFeedMetadataProvider,
+{
+    match PriceFeedMetadataProvider::get_price_feeds_metadata(state, query, asset_type).await {
+        Ok(price_feeds_with_update_data) => Ok(price_feeds_with_update_data),
+        Err(e) => Err(e),
+    }
+}
+
+
+#[async_trait::async_trait]
+pub trait PriceFeedMetadataProvider {
+    async fn get_price_feeds_metadata(
+        &self,
+        query: Option<String>,
+        asset_type: Option<AssetType>,
+    ) -> Result<Vec<PriceFeedMetadata>>;
+}
+
+#[async_trait::async_trait]
+impl PriceFeedMetadataProvider for crate::state::State {
+    async fn get_price_feeds_metadata(
+        &self,
+        query: Option<String>,
+        asset_type: Option<AssetType>,
+    ) -> Result<Vec<PriceFeedMetadata>> {
+        let mut price_feeds_metadata = {
+            let feeds = retrieve_price_feeds_metadata(&self).await?;
+            if feeds.is_empty() {
+                // If the result is an empty Vec, fetch and store new price feeds
+                fetch_and_store_price_feeds_metadata(&self).await?
```

Comment:
This would be classed as `surprising` behaviour to the method caller, as the method implies it just fetches data. We shouldn't mutate within a getter, It's better to return an error here. I think you've done this because on first launch there won't be any state yet, but the better way to do this would be to explicitly fetch this state on launch somewhere before this code is called and rely on making sure `/ready` is aware.